### PR TITLE
UPnP support

### PR DESCRIPTION
**Pull request by s-vincent**
This PR adds:
- miniupnpc 'attendee' dependency in teapot party;
- miniupnpcplus library based on miniupnpc;
- integration of miniupnpcplus in libfscp;
- adds UPnP option in configuration file;
- update vcxproj;
- update docker/vagrant/travis CI.

Note that I put the UPnP stuff in libfscp but I don't mind moving it elsewhere if needed.

Limitation:
- Register same external port as local port;
- No retry for the moment if port is already taken.

Feel free to review, test and give feedbacks :).

### CODE REVIEW

**Comment by leggewie** [comment]
Thank you for this work.

I feel that freelan is already bigger than it should be (I would eventually like to run it on embedded machines, seems natural territory for VPN software).  As such, it would be nice if the feature could be turned off at compilation time.

**Comment by ereOn** [comment]
This is amazing work @s-vincent ! I've made a couple of comments that you might want to check out. Nothing too major, except for a bug waiting to happen in regards to the fact that the UPNP managing class can be incorrectly copied.

I second @leggewie's remark regarding the ability to disable the feature at compile time for binary size reasons.

**Comment by s-vincent** [comment]
@ereOn same as other PR, I don't find the comments in the "Files changed" tab :(.

**Comment by leggewie** [comment]
@s-vincent I don't understand your last comment. Can you explain? Apologies for being dense.

**Comment by s-vincent** [comment]
PR updated to be able to compile or not upnp support (scons --upnp=yes|no) and makes miniupnpcplus::upnp_device non-copyable.

**Comment by ereOn** [comment]
I just realized one has to also click the "Submit review" once the review comments are done... silly me.

File: samples/miniupnpcplus/mapping/mapping.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	static_cast<void>(argv);
+
+	try
+	{
+		miniupnpcplus::upnp_device dev(2000);
+		std::string tmp;
+
+		std::cout << "UPnP device created" << std::endl;
+
+		dev.register_port_mapping(miniupnpcplus::TCP, "12000", "12000",
```
Comment (s-vincent) [inline samples/miniupnpcplus/mapping/mapping.cpp]: Currently all the reamining registered port are unregistered in upnp_device destructor.

Comment (ereOn) [inline samples/miniupnpcplus/mapping/mapping.cpp]: Yeah, that's great.

I was more thinking about an API that would be used like:

```
{
  auto registration = dev.get_port_mapping_registration(...);
  // ...
} // Here registration goes out of scope and so the associated registration gets deleted automatically.
```

But this is not mandatory.

File: libs/miniupnpcplus/include/miniupnpcplus/miniupnpcplus.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * version.  If you delete this exception statement from all source
+ * files in the program, then also delete it here.
+ *
+ * If you intend to use libfscp in a commercial software, please
+ * contact me : we may arrange this for a small fee or no fee at all,
+ * depending on the nature of your project.
+ */
+
+/**
+ * \file executeplus.hpp
```
Comment (s-vincent) [inline libs/miniupnpcplus/include/miniupnpcplus/miniupnpcplus.hpp]: Done.

File: libs/miniupnpcplus/src/upnp_device.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  }
+
+  void upnp_device::unregister_port_mapping(enum transport_protocol protocol,
+      const std::string& external_port)
+  {
+    int error = UPNP_DeletePortMapping(m_upnp_urls.controlURL,
+        m_upnp_data.first.servicetype, external_port.c_str(),
+        (protocol == TCP ? "TCP" : "UDP"), nullptr);
+
+    if(!error)
```
Comment (s-vincent) [inline libs/miniupnpcplus/src/upnp_device.cpp]: Done.

File: libs/miniupnpcplus/include/miniupnpcplus/upnp_device.hpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+       * \brief Returns LAN IP address.
+       * \return LAN IP address.
+       */
+      std::string get_lan_ip() const;
+
+    private:
+      /**
+       * \brief Underlying miniupnpc library pointer.
+       */
+      struct UPNPDev* m_upnp;
```
Comment (s-vincent) [inline libs/miniupnpcplus/include/miniupnpcplus/upnp_device.hpp]: Done.

File: samples/miniupnpcplus/mapping/mapping.cpp (hunk 1 above)
Comment (s-vincent) [inline samples/miniupnpcplus/mapping/mapping.cpp]: Maybe we can first merge this feature and enhance it later (register an RFE issue for that) ?

**Comment by s-vincent** [comment]
Except for the proposed RAII-designed stuff to register port, every other asked changes are done.

File: samples/miniupnpcplus/mapping/mapping.cpp (hunk 1 above)
Comment (ereOn) [inline samples/miniupnpcplus/mapping/mapping.cpp]: Sure thing. I was just asking ;)

**Pull request merged**